[ICE]: `Scalar` layout for non-primitive non-enum type unsafe

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
#![feature(unsafe_binders)]

#[derive(Copy, Clone)]
struct Adt<'a> {
    a: &'a String,
}
const None: Option<unsafe<> Option<unsafe<'a> Adt<'a>>> = None;
fn main() {
    match None {
        _ => {}
    };
}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.96.0-nightly (80d0e4be6 2026-03-25)
binary: rustc
commit-hash: 80d0e4be6f15899649ba31669077c59a986f96cc
commit-date: 2026-03-25
host: x86_64-unknown-linux-gnu
release: 1.96.0-nightly
LLVM version: 22.1.2
```

### Error output

```
warning: the feature `unsafe_binders` is incomplete and may not be safe to use and/or cause compiler crashes
 --> a.rs:1:12
  |
1 | #![feature(unsafe_binders)]
  |            ^^^^^^^^^^^^^^
  |
  = note: see issue #130516 <https://github.com/rust-lang/rust/issues/130516> for more information
  = note: `#[warn(incomplete_features)]` on by default

error[E0277]: the trait bound `unsafe<'a> Adt<'a>: Copy` is not satisfied
 --> a.rs:7:13
  |
7 | const None: Option<unsafe<> Option<unsafe<'a> Adt<'a>>> = None;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Copy` is not implemented for `unsafe<'a> Adt<'a>`
  |
  = note: required for `Option<unsafe<'a> Adt<'a>>` to implement `Copy`

error[E0277]: the trait bound `unsafe<'a> Adt<'a>: Copy` is not satisfied
 --> a.rs:7:59
  |
7 | const None: Option<unsafe<> Option<unsafe<'a> Adt<'a>>> = None;
  |                                                           ^^^^ the trait `Copy` is not implemented for `unsafe<'a> Adt<'a>`
  |
  = note: required for `Option<unsafe<'a> Adt<'a>>` to implement `Copy`
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' (3890271) panicked at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_ty_utils/src/layout/invariant.rs:120:25:
`Scalar` layout for non-primitive non-enum type unsafe<> Option<unsafe<'a> Adt<'a>>
stack backtrace:
   0:     0x7f400d9437b0 - std[4f619bdb2e663cc]::backtrace_rs::backtrace::libunwind::trace
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/../../backtrace/src/backtrace/libunwind.rs:117:9
   1:     0x7f400d9437b0 - std[4f619bdb2e663cc]::backtrace_rs::backtrace::trace_unsynchronized::<std[4f619bdb2e663cc]::sys::backtrace::_print_fmt::{closure#1}>
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/../../backtrace/src/backtrace/mod.rs:66:14
   2:     0x7f400d9437b0 - std[4f619bdb2e663cc]::sys::backtrace::_print_fmt
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/sys/backtrace.rs:74:9
   3:     0x7f400d9437b0 - <<std[4f619bdb2e663cc]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[e58e1890be41adad]::fmt::Display>::fmt
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/sys/backtrace.rs:44:26
   4:     0x7f4008ee6cbe - <core[e58e1890be41adad]::fmt::rt::Argument>::fmt
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/core/src/fmt/rt.rs:152:76
   5:     0x7f4008ee6cbe - core[e58e1890be41adad]::fmt::write
   6:     0x7f400d95b8dc - std[4f619bdb2e663cc]::io::default_write_fmt::<std[4f619bdb2e663cc]::sys::stdio::unix::Stderr>
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/io/mod.rs:639:11
   7:     0x7f400d95b8dc - <std[4f619bdb2e663cc]::sys::stdio::unix::Stderr as std[4f619bdb2e663cc]::io::Write>::write_fmt
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/io/mod.rs:1994:13
   8:     0x7f400d908286 - <std[4f619bdb2e663cc]::sys::backtrace::BacktraceLock>::print
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/sys/backtrace.rs:47:9
   9:     0x7f400d908286 - std[4f619bdb2e663cc]::panicking::default_hook::{closure#0}
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/panicking.rs:292:27
  10:     0x7f400d9324a1 - std[4f619bdb2e663cc]::panicking::default_hook
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/panicking.rs:319:9
  11:     0x7f400a3d4c27 - <alloc[93d8057a61472300]::boxed::Box<dyn for<'a, 'b> core[e58e1890be41adad]::ops::function::Fn<(&'a std[4f619bdb2e663cc]::panic::PanicHookInfo<'b>,), Output = ()> + core[e58e1890be41adad]::marker::Send + core[e58e1890be41adad]::marker::Sync> as core[e58e1890be41adad]::ops::function::Fn<(&std[4f619bdb2e663cc]::panic::PanicHookInfo,)>>::call
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/alloc/src/boxed.rs:2254:9
  12:     0x7f400a3d4c27 - rustc_driver_impl[c9be22246af3610]::install_ice_hook::{closure#1}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_driver_impl/src/lib.rs:1483:17
  13:     0x7f400a3d4c27 - <alloc[93d8057a61472300]::boxed::Box<rustc_driver_impl[c9be22246af3610]::install_ice_hook::{closure#1}> as core[e58e1890be41adad]::ops::function::Fn<(&dyn for<'a, 'b> core[e58e1890be41adad]::ops::function::Fn<(&'a std[4f619bdb2e663cc]::panic::PanicHookInfo<'b>,), Output = ()> + core[e58e1890be41adad]::marker::Send + core[e58e1890be41adad]::marker::Sync, &std[4f619bdb2e663cc]::panic::PanicHookInfo)>>::call
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/alloc/src/boxed.rs:2254:9
  14:     0x7f400a3d4c27 - std[4f619bdb2e663cc]::panicking::update_hook::<alloc[93d8057a61472300]::boxed::Box<rustc_driver_impl[c9be22246af3610]::install_ice_hook::{closure#1}>>::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/panicking.rs:235:47
  15:     0x7f400d93283a - <alloc[93d8057a61472300]::boxed::Box<dyn for<'a, 'b> core[e58e1890be41adad]::ops::function::Fn<(&'a std[4f619bdb2e663cc]::panic::PanicHookInfo<'b>,), Output = ()> + core[e58e1890be41adad]::marker::Send + core[e58e1890be41adad]::marker::Sync> as core[e58e1890be41adad]::ops::function::Fn<(&std[4f619bdb2e663cc]::panic::PanicHookInfo,)>>::call
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/alloc/src/boxed.rs:2254:9
  16:     0x7f400d93283a - std[4f619bdb2e663cc]::panicking::panic_with_hook
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/panicking.rs:833:13
  17:     0x7f400d908348 - std[4f619bdb2e663cc]::panicking::panic_handler::{closure#0}
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/panicking.rs:698:13
  18:     0x7f400d8f70f9 - std[4f619bdb2e663cc]::sys::backtrace::__rust_end_short_backtrace::<std[4f619bdb2e663cc]::panicking::panic_handler::{closure#0}, !>
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/sys/backtrace.rs:182:18
  19:     0x7f400d90a29d - __rustc[bc94a6416cf16b4b]::rust_begin_unwind
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/panicking.rs:689:5
  20:     0x7f4008ee791c - core[e58e1890be41adad]::panicking::panic_fmt
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/core/src/panicking.rs:80:14
  21:     0x7f400d7ec598 - rustc_ty_utils[a8e4f9e9a7aab2c]::layout::layout_of
  22:     0x7f400c98f4f1 - rustc_query_impl[f274b0df531ec29f]::query_impl::layout_of::invoke_provider_fn::__rust_begin_short_backtrace
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_query_impl/src/query_impl.rs:116:46
  23:     0x7f400c747560 - rustc_query_impl[f274b0df531ec29f]::execution::execute_job_non_incr::<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_query_impl/src/execution.rs:395:59
  24:     0x7f400c747560 - rustc_middle[e976332cbd676520]::ty::context::tls::enter_context::<rustc_query_impl[f274b0df531ec29f]::execution::execute_job_non_incr<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_middle/src/ty/context/tls.rs:56:9
  25:     0x7f400c747560 - <std[4f619bdb2e663cc]::thread::local::LocalKey<core[e58e1890be41adad]::cell::Cell<*const ()>>>::try_with::<rustc_middle[e976332cbd676520]::ty::context::tls::enter_context<rustc_query_impl[f274b0df531ec29f]::execution::execute_job_non_incr<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>::{closure#0}, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/thread/local.rs:462:12
  26:     0x7f400c747560 - <std[4f619bdb2e663cc]::thread::local::LocalKey<core[e58e1890be41adad]::cell::Cell<*const ()>>>::with::<rustc_middle[e976332cbd676520]::ty::context::tls::enter_context<rustc_query_impl[f274b0df531ec29f]::execution::execute_job_non_incr<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>::{closure#0}, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/thread/local.rs:426:20
  27:     0x7f400c747560 - rustc_middle[e976332cbd676520]::ty::context::tls::enter_context::<rustc_query_impl[f274b0df531ec29f]::execution::execute_job_non_incr<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_middle/src/ty/context/tls.rs:53:9
  28:     0x7f400c747560 - rustc_query_impl[f274b0df531ec29f]::plumbing::start_query::<rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>, rustc_query_impl[f274b0df531ec29f]::execution::execute_job_non_incr<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}>::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_query_impl/src/plumbing.rs:74:9
  29:     0x7f400c747560 - rustc_middle[e976332cbd676520]::ty::context::tls::with_context::<rustc_query_impl[f274b0df531ec29f]::plumbing::start_query<rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>, rustc_query_impl[f274b0df531ec29f]::execution::execute_job_non_incr<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}>::{closure#0}, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_middle/src/ty/context/tls.rs:86:36
  30:     0x7f400c747560 - rustc_middle[e976332cbd676520]::ty::context::tls::with_context_opt::<rustc_middle[e976332cbd676520]::ty::context::tls::with_context<rustc_query_impl[f274b0df531ec29f]::plumbing::start_query<rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>, rustc_query_impl[f274b0df531ec29f]::execution::execute_job_non_incr<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}>::{closure#0}, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>::{closure#0}, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_middle/src/ty/context/tls.rs:75:18
  31:     0x7f400c747560 - rustc_middle[e976332cbd676520]::ty::context::tls::with_context::<rustc_query_impl[f274b0df531ec29f]::plumbing::start_query<rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>, rustc_query_impl[f274b0df531ec29f]::execution::execute_job_non_incr<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}>::{closure#0}, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_middle/src/ty/context/tls.rs:86:5
  32:     0x7f400c747560 - rustc_query_impl[f274b0df531ec29f]::plumbing::start_query::<rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>, rustc_query_impl[f274b0df531ec29f]::execution::execute_job_non_incr<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_query_impl/src/plumbing.rs:61:5
  33:     0x7f400c747560 - rustc_query_impl[f274b0df531ec29f]::execution::execute_job_non_incr::<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_query_impl/src/execution.rs:395:17
  34:     0x7f400c747560 - rustc_query_impl[f274b0df531ec29f]::execution::try_execute_query::<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>, false>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_query_impl/src/execution.rs:303:17
  35:     0x7f400c98f894 - rustc_query_impl[f274b0df531ec29f]::execution::execute_query_non_incr_inner::<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_query_impl/src/execution.rs:596:32
  36:     0x7f400c98f894 - stacker[ad9f6edc0433b920]::maybe_grow::<rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>, rustc_query_impl[f274b0df531ec29f]::execution::execute_query_non_incr_inner<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}>
                               at /rust/deps/stacker-0.1.21/src/lib.rs:57:9
  37:     0x7f400c98f894 - rustc_data_structures[97dd880033bbc9ee]::stack::ensure_sufficient_stack::<rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>, rustc_query_impl[f274b0df531ec29f]::execution::execute_query_non_incr_inner<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>::{closure#0}>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_data_structures/src/stack.rs:21:5
  38:     0x7f400c98f894 - rustc_query_impl[f274b0df531ec29f]::execution::execute_query_non_incr_inner::<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_query_impl/src/execution.rs:596:5
  39:     0x7f400c98f894 - rustc_query_impl[f274b0df531ec29f]::query_impl::layout_of::execute_query_non_incr::__rust_end_short_backtrace
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_query_impl/src/query_impl.rs:80:30
  40:     0x7f400d865717 - rustc_middle[e976332cbd676520]::query::inner::query_get_at::<rustc_middle[e976332cbd676520]::query::caches::DefaultCache<rustc_middle[e976332cbd676520]::ty::PseudoCanonicalInput<rustc_middle[e976332cbd676520]::ty::Ty>, rustc_middle[e976332cbd676520]::query::erase::ErasedData<[u8; 16usize]>>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_middle/src/query/inner.rs:45:17
  41:     0x7f400d865717 - <rustc_middle[e976332cbd676520]::query::plumbing::TyCtxtAt>::layout_of
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_middle/src/query/plumbing.rs:571:46
  42:     0x7f400d865717 - <rustc_middle[e976332cbd676520]::ty::layout::LayoutCx as rustc_middle[e976332cbd676520]::ty::layout::LayoutOf>::spanned_layout_of
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_middle/src/ty/layout.rs:704:17
  43:     0x7f400d7f59bb - <rustc_middle[e976332cbd676520]::ty::layout::LayoutCx as rustc_middle[e976332cbd676520]::ty::layout::LayoutOf>::layout_of
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_middle/src/ty/layout.rs:691:14
  44:     0x7f400d7f59bb - rustc_ty_utils[a8e4f9e9a7aab2c]::layout::layout_of_uncached::{closure#23}::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_ty_utils/src/layout.rs:653:41
  45:     0x7f400d7f59bb - core[e58e1890be41adad]::iter::adapters::map::map_try_fold::<&rustc_middle[e976332cbd676520]::ty::FieldDef, core[e58e1890be41adad]::result::Result<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>, &rustc_middle[e976332cbd676520]::ty::layout::LayoutError>, (), core[e58e1890be41adad]::ops::control_flow::ControlFlow<core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>>, rustc_ty_utils[a8e4f9e9a7aab2c]::layout::layout_of_uncached::{closure#23}::{closure#0}, <core[e58e1890be41adad]::iter::adapters::GenericShunt<core[e58e1890be41adad]::iter::adapters::by_ref_sized::ByRefSized<core[e58e1890be41adad]::iter::adapters::map::Map<core[e58e1890be41adad]::slice::iter::Iter<rustc_middle[e976332cbd676520]::ty::FieldDef>, rustc_ty_utils[a8e4f9e9a7aab2c]::layout::layout_of_uncached::{closure#23}::{closure#0}>>, core[e58e1890be41adad]::result::Result<core[e58e1890be41adad]::convert::Infallible, &rustc_middle[e976332cbd676520]::ty::layout::LayoutError>> as core[e58e1890be41adad]::iter::traits::iterator::Iterator>::try_fold<(), core[e58e1890be41adad]::iter::traits::iterator::Iterator::try_for_each::call<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>::Break>::{closure#0}, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>>::{closure#0}>::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/core/src/iter/adapters/map.rs:95:28
  46:     0x7f400d7f59bb - <core[e58e1890be41adad]::slice::iter::Iter<rustc_middle[e976332cbd676520]::ty::FieldDef> as core[e58e1890be41adad]::iter::traits::iterator::Iterator>::try_fold::<(), core[e58e1890be41adad]::iter::adapters::map::map_try_fold<&rustc_middle[e976332cbd676520]::ty::FieldDef, core[e58e1890be41adad]::result::Result<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>, &rustc_middle[e976332cbd676520]::ty::layout::LayoutError>, (), core[e58e1890be41adad]::ops::control_flow::ControlFlow<core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>>, rustc_ty_utils[a8e4f9e9a7aab2c]::layout::layout_of_uncached::{closure#23}::{closure#0}, <core[e58e1890be41adad]::iter::adapters::GenericShunt<core[e58e1890be41adad]::iter::adapters::by_ref_sized::ByRefSized<core[e58e1890be41adad]::iter::adapters::map::Map<core[e58e1890be41adad]::slice::iter::Iter<rustc_middle[e976332cbd676520]::ty::FieldDef>, rustc_ty_utils[a8e4f9e9a7aab2c]::layout::layout_of_uncached::{closure#23}::{closure#0}>>, core[e58e1890be41adad]::result::Result<core[e58e1890be41adad]::convert::Infallible, &rustc_middle[e976332cbd676520]::ty::layout::LayoutError>> as core[e58e1890be41adad]::iter::traits::iterator::Iterator>::try_fold<(), core[e58e1890be41adad]::iter::traits::iterator::Iterator::try_for_each::call<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>::Break>::{closure#0}, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>>::{closure#0}>::{closure#0}, core[e58e1890be41adad]::ops::control_flow::ControlFlow<core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/core/src/iter/traits/iterator.rs:2511:21
  47:     0x7f400d7f59bb - <core[e58e1890be41adad]::iter::adapters::map::Map<core[e58e1890be41adad]::slice::iter::Iter<rustc_middle[e976332cbd676520]::ty::FieldDef>, rustc_ty_utils[a8e4f9e9a7aab2c]::layout::layout_of_uncached::{closure#23}::{closure#0}> as core[e58e1890be41adad]::iter::traits::iterator::Iterator>::try_fold::<(), <core[e58e1890be41adad]::iter::adapters::GenericShunt<core[e58e1890be41adad]::iter::adapters::by_ref_sized::ByRefSized<core[e58e1890be41adad]::iter::adapters::map::Map<core[e58e1890be41adad]::slice::iter::Iter<rustc_middle[e976332cbd676520]::ty::FieldDef>, rustc_ty_utils[a8e4f9e9a7aab2c]::layout::layout_of_uncached::{closure#23}::{closure#0}>>, core[e58e1890be41adad]::result::Result<core[e58e1890be41adad]::convert::Infallible, &rustc_middle[e976332cbd676520]::ty::layout::LayoutError>> as core[e58e1890be41adad]::iter::traits::iterator::Iterator>::try_fold<(), core[e58e1890be41adad]::iter::traits::iterator::Iterator::try_for_each::call<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>::Break>::{closure#0}, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>>::{closure#0}, core[e58e1890be41adad]::ops::control_flow::ControlFlow<core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/core/src/iter/adapters/map.rs:121:19
  48:     0x7f400d7f59bb - <core[e58e1890be41adad]::iter::adapters::by_ref_sized::ByRefSized<core[e58e1890be41adad]::iter::adapters::map::Map<core[e58e1890be41adad]::slice::iter::Iter<rustc_middle[e976332cbd676520]::ty::FieldDef>, rustc_ty_utils[a8e4f9e9a7aab2c]::layout::layout_of_uncached::{closure#23}::{closure#0}>> as core[e58e1890be41adad]::iter::traits::iterator::Iterator>::try_fold::<(), <core[e58e1890be41adad]::iter::adapters::GenericShunt<core[e58e1890be41adad]::iter::adapters::by_ref_sized::ByRefSized<core[e58e1890be41adad]::iter::adapters::map::Map<core[e58e1890be41adad]::slice::iter::Iter<rustc_middle[e976332cbd676520]::ty::FieldDef>, rustc_ty_utils[a8e4f9e9a7aab2c]::layout::layout_of_uncached::{closure#23}::{closure#0}>>, core[e58e1890be41adad]::result::Result<core[e58e1890be41adad]::convert::Infallible, &rustc_middle[e976332cbd676520]::ty::layout::LayoutError>> as core[e58e1890be41adad]::iter::traits::iterator::Iterator>::try_fold<(), core[e58e1890be41adad]::iter::traits::iterator::Iterator::try_for_each::call<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>::Break>::{closure#0}, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>>::{closure#0}, core[e58e1890be41adad]::ops::control_flow::ControlFlow<core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/core/src/iter/adapters/by_ref_sized.rs:54:9
  49:     0x7f400d7f59bb - <core[e58e1890be41adad]::iter::adapters::GenericShunt<core[e58e1890be41adad]::iter::adapters::by_ref_sized::ByRefSized<core[e58e1890be41adad]::iter::adapters::map::Map<core[e58e1890be41adad]::slice::iter::Iter<rustc_middle[e976332cbd676520]::ty::FieldDef>, rustc_ty_utils[a8e4f9e9a7aab2c]::layout::layout_of_uncached::{closure#23}::{closure#0}>>, core[e58e1890be41adad]::result::Result<core[e58e1890be41adad]::convert::Infallible, &rustc_middle[e976332cbd676520]::ty::layout::LayoutError>> as core[e58e1890be41adad]::iter::traits::iterator::Iterator>::try_fold::<(), core[e58e1890be41adad]::iter::traits::iterator::Iterator::try_for_each::call<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>::Break>::{closure#0}, core[e58e1890be41adad]::ops::control_flow::ControlFlow<rustc_abi[e2095b3bdd0c3997]::layout::ty::TyAndLayout<rustc_middle[e976332cbd676520]::ty::Ty>>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/core/src/iter/adapters/mod.rs:195:14


< snip >

 184:     0x7f400a3d39a5 - rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler::<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_interface/src/interface.rs:500:23
 185:     0x7f400a3c3125 - rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals::<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_interface/src/util.rs:203:17
 186:     0x7f400a3c3125 - rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals::<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_interface/src/util.rs:159:24
 187:     0x7f400a3c3125 - <scoped_tls[aa3cf235b5f1b553]::ScopedKey<rustc_span[7d7c86df1226e373]::SessionGlobals>>::set::<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
                               at /rust/deps/scoped-tls-1.0.1/src/lib.rs:137:9
 188:     0x7f400a3c3125 - rustc_span[7d7c86df1226e373]::create_session_globals_then::<(), rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_span/src/lib.rs:153:21
 189:     0x7f400a3c3125 - rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals::<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/compiler/rustc_interface/src/util.rs:155:17
 190:     0x7f400a3c3125 - std[4f619bdb2e663cc]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/sys/backtrace.rs:166:18
 191:     0x7f400a3da5d3 - std[4f619bdb2e663cc]::thread::lifecycle::spawn_unchecked::<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/thread/lifecycle.rs:91:13
 192:     0x7f400a3da5d3 - <core[e58e1890be41adad]::panic::unwind_safe::AssertUnwindSafe<std[4f619bdb2e663cc]::thread::lifecycle::spawn_unchecked<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}> as core[e58e1890be41adad]::ops::function::FnOnce<()>>::call_once
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/core/src/panic/unwind_safe.rs:275:9
 193:     0x7f400a3da5d3 - std[4f619bdb2e663cc]::panicking::catch_unwind::do_call::<core[e58e1890be41adad]::panic::unwind_safe::AssertUnwindSafe<std[4f619bdb2e663cc]::thread::lifecycle::spawn_unchecked<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}>, ()>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/panicking.rs:581:40
 194:     0x7f400a3da5d3 - std[4f619bdb2e663cc]::panicking::catch_unwind::<(), core[e58e1890be41adad]::panic::unwind_safe::AssertUnwindSafe<std[4f619bdb2e663cc]::thread::lifecycle::spawn_unchecked<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}>>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/panicking.rs:544:19
 195:     0x7f400a3da5d3 - std[4f619bdb2e663cc]::panic::catch_unwind::<core[e58e1890be41adad]::panic::unwind_safe::AssertUnwindSafe<std[4f619bdb2e663cc]::thread::lifecycle::spawn_unchecked<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}>, ()>
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/panic.rs:359:14
 196:     0x7f400a3da5d3 - std[4f619bdb2e663cc]::thread::lifecycle::spawn_unchecked::<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/thread/lifecycle.rs:89:26
 197:     0x7f400a3da5d3 - <std[4f619bdb2e663cc]::thread::lifecycle::spawn_unchecked<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_with_globals<rustc_interface[8ee6c83dc6f50ddf]::util::run_in_thread_pool_with_globals<rustc_interface[8ee6c83dc6f50ddf]::interface::run_compiler<(), rustc_driver_impl[c9be22246af3610]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[e58e1890be41adad]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
                               at /rustc-dev/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/core/src/ops/function.rs:250:5
 198:     0x7f400d94063f - <alloc[93d8057a61472300]::boxed::Box<dyn core[e58e1890be41adad]::ops::function::FnOnce<(), Output = ()> + core[e58e1890be41adad]::marker::Send> as core[e58e1890be41adad]::ops::function::FnOnce<()>>::call_once
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/alloc/src/boxed.rs:2240:9
 199:     0x7f400d94063f - <std[4f619bdb2e663cc]::sys::thread::unix::Thread>::new::thread_start
                               at /rustc/3a933e5bf5b695faf9e09d47a263a20c6b85c26a/library/std/src/sys/thread/unix.rs:118:17
 200:     0x7f40076a597a - <unknown>
 201:     0x7f40077292bc - <unknown>
 202:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/tmp/im3/rustc-ice-2026-03-26T14_55_56-3890269.txt` to your bug report

note: rustc 1.96.0-nightly (3a933e5bf 2026-03-26) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [layout_of] computing layout of `unsafe<> core::option::Option<unsafe<'a> Adt<'a>>`
#1 [layout_of] computing layout of `core::option::Option<unsafe<> core::option::Option<unsafe<'a> Adt<'a>>>`
... and 2 other queries... use `env RUST_BACKTRACE=1` to see the full query stack
error: aborting due to 2 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0277`.
```

</p>
</details>

--- a/compiler/rustc_ty_utils/src/layout/invariant.rs
+++ b/compiler/rustc_ty_utils/src/layout/invariant.rs
@@ -1,8 +1,8 @@
 use std::assert_matches;
 
 use rustc_abi::{BackendRepr, FieldsShape, Scalar, Size, TagEncoding, Variants};
-use rustc_middle::bug;
 use rustc_middle::ty::layout::{HasTyCtxt, LayoutCx, TyAndLayout};
+use rustc_middle::{bug, ty};
 
 /// Enforce some basic invariants on layouts.
 pub(super) fn layout_sanity_check<'tcx>(cx: &LayoutCx<'tcx>, layout: &TyAndLayout<'tcx>) {
@@ -52,6 +52,14 @@ pub(super) fn layout_sanity_check<'tcx>(cx: &LayoutCx<'tcx>, layout: &TyAndLayou
     }
 
     fn skip_newtypes<'tcx>(cx: &LayoutCx<'tcx>, layout: &TyAndLayout<'tcx>) -> TyAndLayout<'tcx> {
+        match *layout.ty.kind() {
+            ty::UnsafeBinder(bound_ty) => {
+                let ty = cx.tcx().instantiate_bound_regions_with_erased(bound_ty.into());
+                return skip_newtypes(cx, &TyAndLayout { ty, ..*layout });
+            }
+            _ => {}
+        }
+
         if matches!(layout.layout.variants(), Variants::Multiple { .. }) {
             // Definitely not a newtype of anything.
             return *layout;

--- a/tests/ui/unsafe-binders/layout-invariant.rs
+++ b/tests/ui/unsafe-binders/layout-invariant.rs
@@ -1,0 +1,17 @@
+// Regression test for https://github.com/rust-lang/rust/issues/154426
+#![feature(unsafe_binders)]
+
+#[derive(Copy, Clone)]
+struct Adt<'a> {
+    a: &'a String,
+}
+
+const None: Option<unsafe<> Option<unsafe<'a> Adt<'a>>> = None;
+//~^ ERROR the trait bound `unsafe<'a> Adt<'a>: Copy` is not satisfied
+//~| ERROR the trait bound `unsafe<'a> Adt<'a>: Copy` is not satisfied
+
+fn main() {
+    match None {
+        _ => {}
+    };
+}

--- a/tests/ui/unsafe-binders/layout-invariant.stderr
+++ b/tests/ui/unsafe-binders/layout-invariant.stderr
@@ -1,0 +1,19 @@
+error[E0277]: the trait bound `unsafe<'a> Adt<'a>: Copy` is not satisfied
+  --> $DIR/layout-invariant.rs:9:13
+   |
+LL | const None: Option<unsafe<> Option<unsafe<'a> Adt<'a>>> = None;
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Copy` is not implemented for `unsafe<'a> Adt<'a>`
+   |
+   = note: required for `Option<unsafe<'a> Adt<'a>>` to implement `Copy`
+
+error[E0277]: the trait bound `unsafe<'a> Adt<'a>: Copy` is not satisfied
+  --> $DIR/layout-invariant.rs:9:59
+   |
+LL | const None: Option<unsafe<> Option<unsafe<'a> Adt<'a>>> = None;
+   |                                                           ^^^^ the trait `Copy` is not implemented for `unsafe<'a> Adt<'a>`
+   |
+   = note: required for `Option<unsafe<'a> Adt<'a>>` to implement `Copy`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0277`.

